### PR TITLE
Mailsender

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
Port mail sending functionality from codenvy.

### CODE REVIEW

**Review by vparfonov — APPROVED** [review]

File: core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/MailSessionProvider.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import javax.mail.PasswordAuthentication;
+import javax.mail.Session;
+import org.eclipse.che.inject.ConfigurationProperties;
+
+/**
+ * Provider of {@link Session} Configuration can be injected from container with help of {@link
+ * ConfigurationProperties} class. In this case all properties that starts with 'mail.' will be used
+ * to create {@link Session}.
+ */
+public class MailSessionProvider implements Provider<Session> {
```
Comment (sleshchenko) [inline core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/MailSessionProvider.java]: Should it be marked with Singleton annotation?

File: core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/deploy/MailModule.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.mail.deploy;
+
+import com.google.inject.AbstractModule;
+import org.eclipse.che.mail.MailSender;
+import org.eclipse.che.mail.MailSessionProvider;
+
+public class MailModule extends AbstractModule {
```
Comment (sleshchenko) [inline core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/deploy/MailModule.java]: I think you can just remove this MailModule =)

File: pom.xml [hunk 3]
```diff
@@ -64,6 +65,13 @@
                 <version>0.7.0</version>
             </dependency>
             <dependency>
+                <groupId>com.github.kirviq</groupId>
+                <artifactId>dumbster</artifactId>
+                <version>${com.dumbster.version}</version>
+                <scope>test</scope>
+            </dependency>
+
```
Comment (sleshchenko) [inline pom.xml]: Remove this empty space

**Review by sleshchenko — APPROVED** [review]
Other OK for me

File: core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/EmailBean.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+/**
+ * Describing e-mail properties.
+ *
+ * @author Igor Vinokur
+ * @author Alexander Garagatyi
+ */
+public class EmailBean {
+  private String from;
+  private String to;
+  private String replyTo;
```
Comment (garagatyi) [inline core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/EmailBean.java:26]: Do you know whether it supports multiple recipients?

Comment (skabashnyuk) [reply core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/EmailBean.java:26]: I don't

File: core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/MailSessionProvider.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.mail.Authenticator;
+import javax.mail.PasswordAuthentication;
+import javax.mail.Session;
+import org.eclipse.che.inject.ConfigurationProperties;
+
+/**
+ * Provider of {@link Session} Configuration can be injected from container with help of {@link
```
Comment (garagatyi) [inline core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/MailSessionProvider.java]: This comment should be put to the appropriate constructor instead. This explanation is not related to the second constructor

File: core/commons/che-core-commons-mail/src/test/java/org/eclipse/che/mail/MailSenderTest.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    assertTrue(email.getBody().contains(body));
+    if (attachmentFileName != null && attachmentContentID != null) {
+      assertTrue(email.getBody().contains("filename=" + attachmentFileName));
+      assertTrue(email.getBody().contains("Content-ID: <" + attachmentContentID + ">"));
+    }
+  }
+
+  @BeforeMethod
+  public void setup(ITestContext context) throws IOException {
+    //      mailSender = new MailSender("/mail-configuration.properties");
```
Comment (garagatyi) [inline core/commons/che-core-commons-mail/src/test/java/org/eclipse/che/mail/MailSenderTest.java]: Can you remove this line?

File: core/commons/che-core-commons-mail/src/test/java/org/eclipse/che/mail/MailSenderTest.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    server = SimpleSmtpServer.start(SimpleSmtpServer.AUTO_SMTP_PORT);
+
+    Map<String, String> mailConfiguration =
+        ImmutableMap.of(
+            "mail.smtp.host",
+            "localhost",
```
Comment (garagatyi) [inline core/commons/che-core-commons-mail/src/test/java/org/eclipse/che/mail/MailSenderTest.java:65]: It looks like an integration test. WDYT?

**Comment by garagatyi** [comment]
Never mind, I found it in tests

**Review by garagatyi — COMMENTED** [review]
Looks good!
Can you take a look at my inline comments?

File: core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/MailSender.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+  @Inject
+  public MailSender(MailSessionProvider mailSessionProvider) {
+    this.mailSessionProvider = mailSessionProvider;
+    this.executor =
+        newFixedThreadPool(
+            2 * Runtime.getRuntime().availableProcessors(),
+            new ThreadFactoryBuilder()
+                .setNameFormat("MailNotificationsPool-%d")
+                .setDaemon(false)
```
[the remark this answers is not in the document]
Comment (skabashnyuk) [reply core/commons/che-core-commons-mail/src/main/java/org/eclipse/che/mail/MailSender.java:59]: done

File: core/commons/che-core-commons-mail/src/test/java/org/eclipse/che/mail/MailSenderTest.java (hunk 7 above)
Comment (skabashnyuk) [reply core/commons/che-core-commons-mail/src/test/java/org/eclipse/che/mail/MailSenderTest.java:65]: yes, very tiny one.

**Review by garagatyi — APPROVED** [review]

File: dockerfiles/init/manifests/che.env [hunk 9]
```diff
@@ -455,6 +455,29 @@ CHE_SINGLE_PORT=false
...
+########################################################################################
+#####                                                                              #####
+#####                                 MAIL SERVER                                  #####
+#####                                                                              #####
+#
+# CHE uses email to notify admins and users of events in the system such as error
+# reports invitations, or workspace unusual activity.
+#
+#CHE_MAIL_HOST=smtp.example.com
+#CHE_MAIL_HOST_PORT=465
```
Comment (riuvshin) [inline dockerfiles/init/manifests/che.env:469]: these lines are commented, i/e it won't be injected, where do you have default values for it? I mean we usually have them in che.properties or hardcoded in code but I don't see any

Comment (skabashnyuk) [reply dockerfiles/init/manifests/che.env:469]: Nowhere . By default, it will fail.

Comment (riuvshin) [reply dockerfiles/init/manifests/che.env:469]: so with these changes che with default settings will fail to start, right? is that acceptable?

Comment (skabashnyuk) [reply dockerfiles/init/manifests/che.env:469]: I would not fail on tomcat start. I think we can uncomment this default values in future if we would like to do so.
